tsl/src/chunk_copy.c

```diff
@@ -624,13 +817,185 @@ chunk_copy(Oid chunk_relid, const char *src_node, const char *dst_node, bool del
+		cc->stage = NULL;
+
+		/* No other sanity checks need to be performed since they were done earlier */
+
+		/* Setup the src_node */
+		cc->src_server =
+			data_node_get_foreign_server(NameStr(cc->fd.source_node_name), ACL_USAGE, true, false);
+		Assert(NULL != cc->src_server);
+
+		/* Setup the dst_node */
+		cc->dst_server =
+			data_node_get_foreign_server(NameStr(cc->fd.dest_node_name), ACL_USAGE, true, false);
+		Assert(NULL != cc->dst_server);
+	}
+
+	MemoryContextSwitchTo(old);
+
+	if (cc == NULL)
+		/* No entry found, long lived context not required */
+		MemoryContextDelete(mcxt);
+
+	return cc;
+}
+
+static void
+chunk_copy_cleanup_internal(ChunkCopy *cc, int stage_idx)
+{
+	bool first = true;
+
+	/* Cleanup each copy stage in a separate transaction */
+	do
+	{
+		StartTransactionCommand();
+
+		cc->stage = &chunk_copy_stages[stage_idx];
+		if (cc->stage->function_cleanup)
+			cc->stage->function_cleanup(cc);
+
+		/* Mark stage as cleaned up and update the catalog */
+		if (!first && stage_idx != 0)
+			chunk_copy_activity_update(cc);
+		else
+			first = false;
+
+		CommitTransactionCommand();
+	} while (--stage_idx >= 0);
+}
+
+void
+chunk_copy_cleanup(const char *operation_id)
+{
+	ChunkCopy *cc;
+	const MemoryContext oldcontext = CurrentMemoryContext;
+	const ChunkCopyStage *stage;
+	bool found = false;
+	int stage_idx;
+
+	if (!superuser())
+		ereport(ERROR,
+				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+				 (errmsg("must be superuser to cleanup a chunk copy operation"))));
+
+	if (dist_util_membership() != DIST_MEMBER_ACCESS_NODE)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("function must be run on the access node only")));
+
+	cc = chunk_copy_activity_get(operation_id);
+
+	if (cc == NULL)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid chunk copy activity identifier. Entry not found")));
+
+	/* Identify the last completed stage for this activity. */
+	stage_idx = 0;
+	for (stage = &chunk_copy_stages[stage_idx]; stage->name != NULL;
+		 stage = &chunk_copy_stages[++stage_idx])
+	{
+		if (namestrcmp(&cc->fd.completed_stage, stage->name) == 0)
+		{
+			found = true;
+			break;
+		}
+	}
+
+	/* should always find an entry, add ereport to quell compiler warning */
+	Assert(found == true);
+	if (!found)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("stage '%s' not found for copy chunk cleanup",
+						NameStr(cc->fd.completed_stage))));
+
+	/* Commit to get out of starting transaction */
+	PopActiveSnapshot();
+	CommitTransactionCommand();
+
+	/* Run the corresponding cleanup steps to roll back the activity. */
+	PG_TRY();
+	{
+		chunk_copy_cleanup_internal(cc, stage_idx);
+	}
+	PG_CATCH();
+	{
+		/* Include chunk copy id to the error message */
+		ErrorData *edata;
+		MemoryContextSwitchTo(oldcontext);
+		edata = CopyErrorData();
+		edata->hint = psprintf("While cleaning up chunk copy operation id: (%s).",
```

Comment:
This should probably also be an errdetail, rather than hint, since it is not suggesting how to fix a problem.

Reply:
Done.